### ISSUE TITLE
[promtail] Regex issue with the pipelineStages YAML value

### ISSUE THREAD

**Issue by adesprez** [grafana/helm-charts#2428]
Given the following values for `.Values.config.snippets.pipelineStages`:
```
config:
[...]
    pipelineStages:
    - multiline:
        firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'
        max_wait_time: 3s
```

Expected result for the promtail `Secret`:
```
scrape_configs:
  # See also https://github.com/grafana/loki/blob/master/production/ksonnet/promtail/scrape_config.libsonnet for reference
  - job_name: kubernetes-pods
    pipeline_stages:
      - cri: {}
      - multiline:
          firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'
          max_wait_time: 3s
```

Actual result for the promtail `Secret`:
```
scrape_configs:
  # See also https://github.com/grafana/loki/blob/master/production/ksonnet/promtail/scrape_config.libsonnet for reference
  - job_name: kubernetes-pods
    pipeline_stages:
      - cri: {}
      - multiline:
          firstline: ^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}
          max_wait_time: 3s
```
(no single quotes `'` enclosing the regex).

As a consequence, Promtail is not processing that `firstline` as a regex. Single quotes are mandatory here.

Code responsible of that, `charts/promtail/values.yaml`:
```
    scrapeConfigs: |
      # See also https://github.com/grafana/loki/blob/master/production/ksonnet/promtail/scrape_config.libsonnet for reference
      - job_name: kubernetes-pods
        pipeline_stages:
          {{- toYaml .Values.config.snippets.pipelineStages | nindent 4 }}
```

The `toYaml` function is stripping the single quotes.

Something that could solve that issue, is to use `tpl` function to render `pipelineStages`, as suggested here by tiagodj: https://github.com/grafana/helm-charts/issues/2412

**Comment by xyfleet** [comment]
I had same issue here. Any progress for this one?

**Comment by adesprez** [comment]
@xyfleet , I haven't found a workaround.
I tried multiple syntaxes to escape characters and/or protect the single quotes. No luck. 
I have parked this, since then.

**Comment by chrizel** [comment]
Having the same problem, I can confirm that the `tpl` function can be used to work around this problem by moving the regex to its own custom values variable and then using it in the `pipelineStages` definition:

```yaml
  - match:
      selector: '{name="foobar"}'
      stages:
      - regex:
          expression: {{ .Values.mySpecialRegex }}
```

I have created PR #2575 that makes the `tpl` modification in the promtail helm chart.

**Comment by xyfleet** [comment]
@chrizel Thanks for your update. The new code has been released. I tried to updated to the new release, 6.14.1 and found the secret still does not have single quote. Looks like the tpl did not work as expected. Is there anything something with my setting?

<img width="483" alt="image" src="https://github.com/grafana/helm-charts/assets/113627389/1b0214ed-b46b-42f3-873a-1b8dc2071693">

I am using the same setting as describe before.

```
snippets:
    pipelineStages:
      - multiline:
          firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'
          max_lines: 128
          max_wait_time: 3s
```

**Comment by zanhsieh** [comment]
Folks, can somebody trace down to configmap? Install promtail -> go to promtail container -> cat `/etc/promtail/promtail.yaml`

https://github.com/grafana/helm-charts/blob/d8a64071970b5928cf990dce480925ff6b8edb56/charts/promtail/templates/_pod.tpl#L47
https://github.com/grafana/helm-charts/blob/d8a64071970b5928cf990dce480925ff6b8edb56/charts/promtail/templates/secret.yaml#L17

**Comment by andrejshapal** [comment]
Hello, 
Guys, you broke everything?
Because preveously everything was working fine and not anymore. 
I see exactly all '' are removed from my config after bumping to promtail-6.14.1

**Comment by andrejshapal** [comment]
@chrizel I fixed it by rollback (setting old config manually in my values.yaml snippet). 
I looked into promtail source code and I see yaml parsing and mapping is used. Also, there are errors for every case if decoing failed. 
If something is wrong with '', there is a bug.

**Comment by andrejshapal** [comment]
@chrizel I made tests to confirm issue and did not find anything.
**Here we have logs I am producing:**
<img width="259" alt="image" src="https://github.com/grafana/helm-charts/assets/96792836/b0acaec1-1d8b-4445-be94-9ed0c7ae580c">
**Here is my pipeline config:**
`        snippets:
          pipelineStages:
            - docker: {}
            - multiline:
                firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\\n'
                max_wait_time: 3s
            - regex:
                expression: '^(?P<time>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}) (?P<message>(?s:.*))$'`
The rest is default from chart version targetRevision: 6.6.1.

**My secret content:**
`server:
  log_level: info
  http_listen_port: 3101
  

clients:
  - tenant_id: 1
    url: http://loki-gateway.logging/loki/api/v1/push

positions:
  filename: /run/promtail/positions.yaml

scrape_configs:
  # See also https://github.com/grafana/loki/blob/master/production/ksonnet/promtail/scrape_config.libsonnet for reference
  - job_name: kubernetes-pods
    pipeline_stages:
      - docker: {}
      - multiline:
          firstline: ^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\\n
          max_wait_time: 3s
      - regex:
          expression: ^(?P<time>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}) (?P<message>(?s:.*))$
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - source_labels:
          - __meta_kubernetes_pod_controller_name
        regex: ([0-9a-z-.]+?)(-[0-9a-f]{8,10})?
        action: replace
        target_label: __tmp_controller_name
      - source_labels:
          - __meta_kubernetes_pod_label_app_kubernetes_io_name
          - __meta_kubernetes_pod_label_app
          - __tmp_controller_name
          - __meta_kubernetes_pod_name
        regex: ^;*([^;]+)(;.*)?$
        action: replace
        target_label: app
      - source_labels:
          - __meta_kubernetes_pod_label_app_kubernetes_io_instance
          - __meta_kubernetes_pod_label_release
        regex: ^;*([^;]+)(;.*)?$
        action: replace
        target_label: instance
      - source_labels:
          - __meta_kubernetes_pod_label_app_kubernetes_io_component
          - __meta_kubernetes_pod_label_component
        regex: ^;*([^;]+)(;.*)?$
        action: replace
        target_label: component
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_node_name
        target_label: node_name
      - action: replace
        source_labels:
        - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        replacement: $1
        separator: /
        source_labels:
        - namespace
        - app
        target_label: job
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
        - __meta_kubernetes_pod_container_name
        target_label: container
      - action: replace
        replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
        - __meta_kubernetes_pod_uid
        - __meta_kubernetes_pod_container_name
        target_label: __path__
      - action: replace
        regex: true/(.*)
        replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
        - __meta_kubernetes_pod_annotationpresent_kubernetes_io_config_hash
        - __meta_kubernetes_pod_annotation_kubernetes_io_config_hash
        - __meta_kubernetes_pod_container_name
        target_label: __path__
  
  

limits_config:
`
What I see in grafana:
<img width="831" alt="image" src="https://github.com/grafana/helm-charts/assets/96792836/2994f988-3733-42e0-9516-7176609a3286">

Please, send me your full config (promtail) (where you see an issue). Add how your logs looks in grafana. 
Mention your helm chart version and promtail version.

**Comment by chrizel** [comment]
@andrejshapal I will look into my regex problem when I have more time, but it was in line with what @adesprez wrote.

In the meantime, I have created PR #2584 which reverts the `tpl` change. Sorry for the inconvenience.

**Comment by andrejshapal** [comment]
@chrizel Don't worry. It is ok. 
Thank you for reverting.

**Comment by xyfleet** [comment]
Thanks everyone. Do we know which old version does not have this issue we can rollback? Or any other idea to fix this issue?

**Comment by andrejshapal** [comment]
@xyfleet Which issue? We have two.

**Comment by xyfleet** [comment]
@andrejshapal My issue is that the promtail cannot put multiline logs into one line. As a result, a entire log was divided into multilines on Grafana. Based on the previous discussion, this issue was caused by the missing '' in the `firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'`

**Comment by andrejshapal** [comment]
@xyfleet I don't think this is your issue. 
Few posts above I have shown the test and '' is not deeded. 
Please, send me your full config (promtail) (where you see an issue). Add how your logs looks in grafana.
Mention your helm chart version and promtail version.

**Comment by xyfleet** [comment]
@andrejshapal my loki setting: (I only update the config part, very simple.)

```
config:
  clients:
    - url: http://loki:3100/loki/api/v1/push
  snippets:
    pipelineStages:
      - multiline:
          firstline: '^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'
          max_lines: 128
          max_wait_time: 3s
```

The same log in Pod and in Grafana. From the screenshot, the log in Grafana is divided into several lines with their own timestamp. 
![Screenshot2023_08_11_094345](https://github.com/grafana/helm-charts/assets/113627389/5adbf2e3-f76c-46dc-a9e7-3afed0821456)


Log in Grafana

![Screenshot2023_08_11_095011](https://github.com/grafana/helm-charts/assets/113627389/68e5d95e-31dc-453c-8255-69869093b359)

**Comment by andrejshapal** [comment]
@xyfleet Unfortunately on second screenshot (from grafana) I can't see exactly the first line which regex should catch. It rather important. 
From what I can see without it:

1.  `'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}'` In the log timeframe looks like dddd-dd-dd dd:dd:dd, with regex you try to catch dddd-dd-ddTdd:dd:dd. This will never work.
2. Additionally, as said in https://grafana.com/docs/loki/latest/clients/promtail/stages/multiline/ :
>     # Flag (?s:.*) needs to be set for regex stage to capture full traceback log in the extracted map.
So, basically (if I understand correct), regex stage is compulsory to get the time from multiline. You have to add:
```
- regex:
    expression: '^(?P<time>\[\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}\]) (?P<message>(?s:.*))$'
```

**Comment by xyfleet** [comment]
@andrejshapal 

Based on your suggestion, I updated the Promtail config setting. Can you check?

```
 pipelineStages:
      - multiline:
          firstline: '^\[\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}\]'
          max_lines: 128
          max_wait_time: 3s
      - regex:
          expression: '^(?P<time>\[\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}\]) (?P<message>(?s:.*))$'
VALUES
```

Please see the new screenshots.

log from pod:
![Screenshot2023_08_11_121051](https://github.com/grafana/helm-charts/assets/113627389/9de1f736-da86-4d48-93aa-bcab1618537e)

log from grafana:
![Screenshot2023_08_11_121205](https://github.com/grafana/helm-charts/assets/113627389/0ab49bf9-b43f-414f-a47b-a2686786207d)

**Comment by andrejshapal** [comment]
Not exactly. 
Your time 2023-08-11 18:07:16.189 is not enclosed in [].
Therefore your config should be:
 ```
pipelineStages:
      - multiline:
          firstline: '^\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}' // https://regex101.com/r/DRyksF/1
          max_lines: 128
          max_wait_time: 3s
      - regex:
          expression: '^(?P<time>\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2})(?P<message>(?s:.*))$' // https://regex101.com/r/0a3gUP/1
```

Additionally, I have a suspicion you may need parser before doing multiline:
 ```
pipelineStages:
      - cri: {}
      - multiline:
          firstline: '^\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}'
          max_lines: 128
          max_wait_time: 3s
      - regex:
          expression: '^(?P<time>\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2})(?P<message>(?s:.*))$'
```
I am not really sure. Just think so looking on your latest screenshot. Don't know why, but they have different format with the first one.

**Comment by xyfleet** [comment]
@andrejshapal 
Back to your question "Don't know why, but they have different format with the first one." There are two formats of time. 

![Screenshot2023_08_11_110125](https://github.com/grafana/helm-charts/assets/113627389/8d45d5c7-0de6-40de-b719-9f9e79525b36)
This format comes from my original setting without any `multiline` and `regex`. 


![Screenshot2023_08_11_150448](https://github.com/grafana/helm-charts/assets/113627389/ff68447c-f2fb-44a7-acdc-2c9ad5bcb346)
This format comes from the updated setting with `multiline` and `regex`


```
pipelineStages:
     - cri: {}
     - multiline:
           firstline: '^\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}'
           max_lines: 128
           max_wait_time: 3s
     - regex:
          expression: '^(?P<time>\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2})(?P<message>(?s:.*))$'
```
After several tests, the message still got divided into multilines as you see before.

**Comment by andrejshapal** [comment]
> I thought this is caused by the missing '' in the secret. Isn't it?
There are no issue with missing '' because nothing is missing. 

> This format comes from the updated setting with multiline and regex
I don't think this is possible. 
1. Screenshot - the logs are parsed from cri format
2. Screenshot - logs are not parsed from cri format and regex stage applied
3. Screenshot - logs are not parsed from cri format and regex stage applied
This makes me thing for the first grafana screenshot you did not post the full config of your pipeline stages or you have executed query with some parsers. 
The

**Comment by xyfleet** [comment]
@andrejshapal Thank you so much. After testing with the configuration, I can see the log correctly in Grafana. 
```
pipelineStages:
     - cri: {} // parse log
     - multiline:
           firstline: '^\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2}' // catch first line which starts with timestamp
           max_lines: 128
           max_wait_time: 3s
     - regex:
          expression: '^(?P<time>\d{4}-\d{2}-\d{2} \d{1,2}:\d{2}:\d{2})(?P<message>(?s:.*))$' // after logs combined in one line, retrieve timestamp
```
![Screenshot2023_08_13_103554](https://github.com/grafana/helm-charts/assets/113627389/46a2601a-4905-4aa8-839a-0a57c1e2fdd4)


In my previous tests, I did not add `- cri: {}` to my configuration directly, but this is in the values.yaml file
![Screenshot2023_08_13_103716](https://github.com/grafana/helm-charts/assets/113627389/5c52a614-b3e1-4d8c-b993-446343555b6a)

So, I think the `- cri: {}` should be added to my configuration by default. This is the reason I did not put the configuration in my code as well. But anyway, it works as expected. Thanks.

**Comment by andrejshapal** [comment]
@luizbafilho 
Hello,
What is the problem?
